File: client/lib/auth/auth_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:google_sign_in/google_sign_in.dart';
import 'package:flutter/foundation.dart' show kIsWeb;
import '../config.dart';
import 'session_manager.dart';



/// Result from verify_password or verify_otp
class AuthResult {
  final bool requiresKeySetup;
  final bool isNewUser;
  final String? nextStep;
  final String? message;

  AuthResult({
    required this.requiresKeySetup,
    required this.isNewUser,
    this.nextStep,
    this.message,
  });
}

/// HTTP wrapper that auto-refreshes JWT on 401 responses.
class AuthenticatedClient {
  static Future<http.Response> get(Uri uri) async {
    var token = await SessionManager.getToken();
    var response = await http.get(uri, headers: _headers(token));

    if (response.statusCode == 401) {
      final refreshed = await AuthService.refreshToken();
      if (refreshed) {
        token = await SessionManager.getToken();
        response = await http.get(uri, headers: _headers(token));
      }
    }
    return response;
  }

  static Future<http.Response> post(Uri uri, {Object? body}) async {
    var token = await SessionManager.getToken();
    var response = await http.post(uri, headers: _headers(token), body: body);

    if (response.statusCode == 401) {
      final refreshed = await AuthService.refreshToken();
      if (refreshed) {
        token = await SessionManager.getToken();
        response = await http.post(uri, headers: _headers(token), body: body);
      }
    }
    return response;
  }

  static Future<http.Response> put(Uri uri, {Object? body}) async {
    var token = await SessionManager.getToken();
    var response = await http.put(uri, headers: _headers(token), body: body);

    if (response.statusCode == 401) {
      final refreshed = await AuthService.refreshToken();
      if (refreshed) {
        token = await SessionManager.getToken();
        response = await http.put(uri, headers: _headers(token), body: body);
      }
    }
    return response;
  }

  static Future<http.Response> delete(Uri uri) async {
    var token = await SessionManager.getToken();
    var response = await http.delete(uri, headers: _headers(token));

    if (response.statusCode == 401) {
      final refreshed = await AuthService.refreshToken();
      if (refreshed) {
        token = await SessionManager.getToken();
        response = await http.delete(uri, headers: _headers(token));
      }
    }
    return response;
  }

  static Map<String, String> _headers(String? token) => {
    'Content-Type': 'application/json',
    if (token != null) 'Authorization': 'Bearer $token',
  };
}

class AuthService {
  static final String _baseUrl = AppConfig.apiBaseUrl;

  static final GoogleSignIn _googleSignIn = GoogleSignIn(
    serverClientId: '132565546749-7fjpe9g1ekrtnvjri1dlk2034vmgf2r4.apps.googleusercontent.com',
  );

  /// Sign in with Google
  static Future<AuthResult?> signInWithGoogle() async {
    try {
      final GoogleSignInAccount? googleUser = await _googleSignIn.signIn();
      if (googleUser == null) return null; // User cancelled

      final GoogleSignInAuthentication googleAuth = await googleUser.authentication;
      final String? idToken = googleAuth.idToken;

      if (idToken == null) throw Exception("Could not get ID Token from Google");

      final uri = Uri.parse('$_baseUrl/api/v1/auth/google');
      final response = await http.post(
        uri,
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'id_token': idToken}),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        await SessionManager.saveSession(
          token: data['token'],
          userId: data['user_id'],
          deviceId: data['device_id'],
          refreshToken: data['refresh_token'],
        );
        return AuthResult(
          requiresKeySetup: data['requires_key_setup'] == true,
          isNewUser: data['is_new_user'] == true,
        );
      } else {
        throw Exception(jsonDecode(response.body)['error'] ?? 'Google login failed');
      }
    } catch (e) {
      rethrow;
    }
  }

  /// Step 1: Identify if email/phone exists
  /// Returns { message, next_step, temp_token }
  static Future<Map<String, dynamic>> identify(String identifier) async {
    final uri = Uri.parse('$_baseUrl/api/v1/auth/identify');
    final response = await http.post(
      uri,
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'identifier': identifier}),
    );

    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else if (response.statusCode == 429) {
      throw Exception('Too many attempts. Please try again later.');
    } else {
      throw Exception(jsonDecode(response.body)['error'] ?? 'Identification failed');
    }
  }

  /// Step 2a: Verify Password (login or signup)
  static Future<AuthResult> verifyPassword(String tempToken, String password) async {
    final uri = Uri.parse('$_baseUrl/api/v1/auth/verify_password');
    final response = await http.post(
      uri,
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'temp_token': tempToken,
        'password': password,
      }),
    );

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      
      if (data['next_step'] == 'otp_verify' || data['next_step'] == 'email_collect') {
        return AuthResult(
          requiresKeySetup: false,
          isNewUser: true,
          nextStep: data['next_step'],
          message: data['message'],
        );
      }
      
      await SessionManager.saveSession(
        token: data['token'],
        userId: data['user_id'],
        deviceId: data['device_id'],
        refreshToken: data['refresh_token'],
      );
      return AuthResult(
        requiresKeySetup: data['requires_key_setup'] == true,
        isNewUser: data['is_new_user'] == true,
      );
    } else if (response.statusCode == 422) {
      final data = jsonDecode(response.body);
      final details = (data['details'] as List?)?.join(', ') ?? 'Password too weak';
      throw Exception(details);
    } else {
      throw Exception(jsonDecode(response.body)['error'] ?? 'Invalid credentials');
    }
  }

  /// Step 2b: Verify OTP
  static Future<AuthResult> verifyOtp(String tempToken, String otp) async {
    final uri = Uri.parse('$_baseUrl/api/v1/auth/verify_otp');
    final response = await http.post(
      uri,
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'temp_token': tempToken,
        'otp': otp,
      }),
    );

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      await SessionManager.saveSession(
        token: data['token'],
        userId: data['user_id'],
        deviceId: data['device_id'],
        refreshToken: data['refresh_token'],
      );
      return AuthResult(
        requiresKeySetup: data['requires_key_setup'] == true,
        isNewUser: data['is_new_user'] == true,
      );
    } else {
      final error = jsonDecode(response.body)['error'] ?? 'Invalid OTP';
      throw Exception(error);
    }
  }

  /// Resend OTP to the specified email
  static Future<void> resendOtp(String tempToken, String email) async {
    final uri = Uri.parse('$_baseUrl/api/v1/auth/resend_otp');
    final response = await http.post(
      uri,
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'temp_token': tempToken,
        'email': email,
      }),
    );

    if (response.statusCode == 429) {
      throw Exception('Please wait before requesting a new code.');
    } else if (response.statusCode != 200) {
      throw Exception(jsonDecode(response.body)['error'] ?? 'Failed to resend');
    }
  }

  /// Send OTP for phone number signup (sends to user's provided email)
  static Future<AuthResult> sendPhoneOtp(String tempToken, String email) async {
    final uri = Uri.parse('$_baseUrl/api/v1/auth/send_phone_otp');
    final response = await http.post(
      uri,
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'temp_token': tempToken,
        'email': email,
      }),
    );

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      return AuthResult(
        requiresKeySetup: false,
        isNewUser: true,
        nextStep: data['next_step'],
        message: data['message'],
      );
    } else if (response.statusCode == 422) {
      throw Exception(jsonDecode(response.body)['error'] ?? 'Invalid email');
    } else {
      throw Exception(jsonDecode(response.body)['error'] ?? 'Failed to send verification');
    }
  }

  /// Request a password reset code
  static Future<void> forgotPassword(String email) async {
    final uri = Uri.parse('$_baseUrl/api/v1/auth/forgot_password');
    await http.post(
      uri,
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'email': email}),
    );
    // Always succeeds (anti-enumeration)
  }

  /// Reset password with code
  static Future<void> resetPassword(String email, String code, String newPassword) async {
    final uri = Uri.parse('$_baseUrl/api/v1/auth/reset_password');
    final response = await http.post(
      uri,
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'email': email,
        'code': code,
        'new_password': newPassword,
      }),
    );

    if (response.statusCode == 422) {
      final data = jsonDecode(response.body);
      final details = (data['details'] as List?)?.join(', ') ?? 'Password too weak';
      throw Exception(details);
    } else if (response.statusCode != 200) {
      throw Exception(jsonDecode(response.body)['error'] ?? 'Reset failed');
    }
  }

  /// Silently refresh the access token using the stored refresh token.
  static Future<bool> refreshToken() async {
    final storedRefreshToken = await SessionManager.getRefreshToken();
    if (storedRefreshToken == null) return false;

    final uri = Uri.parse('$_baseUrl/api/v1/auth/refresh');
    try {
      final response = await http.post(
        uri,
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'refresh_token': storedRefreshToken}),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        final userId = await SessionManager.getUserId();
        final deviceId = await SessionManager.getDeviceId();
        await SessionManager.saveSession(
          token: data['token'],
          userId: userId ?? '',
          deviceId: deviceId ?? '',
          refreshToken: data['refresh_token'],
        );
        return true;
      }
    } catch (_) {}
    return false;
  }

  /// Update the user's profile (name and avatar)
  static Future<void> updateProfile({
    required String firstName,
    String? lastName,
    String? imagePath,
  }) async {
    var token = await SessionManager.getToken();
    final uri = Uri.parse('$_baseUrl/api/v1/account/profile');

    Future<http.Response> sendRequest(String? currentToken) async {
      var request = http.MultipartRequest('PUT', uri);
      request.headers['Authorization'] = 'Bearer $currentToken';
      
      request.fields['first_name'] = firstName;
      if (lastName != null && lastName.isNotEmpty) {
        request.fields['last_name'] = lastName;
      }

      if (imagePath != null && imagePath.isNotEmpty) {
        if (kIsWeb) {
          final fileResponse = await http.get(Uri.parse(imagePath));
          request.files.add(http.MultipartFile.fromBytes(
            'avatar',
            fileResponse.bodyBytes,
            filename: 'avatar.jpg',
          ));
        } else {
          request.files.add(await http.MultipartFile.fromPath('avatar', imagePath));
        }
      }

      final streamedResponse = await request.send();
      return await http.Response.fromStream(streamedResponse);
    }

    var response = await sendRequest(token);

    if (response.statusCode == 401) {
      final refreshed = await refreshToken();
      if (refreshed) {
        token = await SessionManager.getToken();
        response = await sendRequest(token);
      }
    }

    if (response.statusCode != 200) {
      throw Exception(jsonDecode(response.body)['error'] ?? 'Failed to update profile');
    }
  }

  /// Fetch the current user's profile
  static Future<Map<String, dynamic>> getProfile() async {
    final uri = Uri.parse('$_baseUrl/api/v1/account/profile');
    final response = await AuthenticatedClient.get(uri);
    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else {
      throw Exception(jsonDecode(response.body)['error'] ?? 'Failed to fetch profile');
    }
  }

  /// Update the user's username and optional PIN
  static Future<void> setUsername({
    required String username,
    String? pin,
  }) async {
    final uri = Uri.parse('$_baseUrl/api/v1/account/username');
    final Map<String, dynamic> body = {'username': username};
    if (pin != null && pin.isNotEmpty) {
      body['pin'] = pin;
    }

    final response = await AuthenticatedClient.put(uri, body: jsonEncode(body));

    if (response.statusCode != 200) {
      final errorBody = jsonDecode(response.body);
      
      if (errorBody['errors'] != null && errorBody['errors'] is Map) {
        final errors = errorBody['errors'] as Map;
        if (errors.containsKey('username')) {
          final msgs = errors['username'] as List;
          throw Exception("Username ${msgs.first}");
        }
      }
      
      throw Exception(errorBody['error'] ?? 'Failed to set username');
    }
  }

  /// Check if a username is available
  static Future<bool> checkUsername(String username) async {
    final uri = Uri.parse('$_baseUrl/api/v1/account/check_username?username=${Uri.encodeComponent(username)}');
    try {
      final response = await http.get(uri);
      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return data['available'] == true;
      }
      return false;
    } catch (_) {
      // In case of network errors, default to false or handle differently.
      // For real-time check, returning false is safer, but returning true might avoid false positives if network flickers.
      // We will let the final submit catch actual errors.
      return true; // Assume available so we don't block typing on network blip
    }
  }
}

File: android_client/app/src/main/java/com/subhankar/aurachat/network/AuthService.kt
package com.subhankar.aurachat.network

import android.util.Log
import com.subhankar.aurachat.data.local.SessionManager
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.IOException
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlinx.coroutines.suspendCancellableCoroutine

/**
 * Authentication result from verify_password or verify_otp.
 * Matches Flutter's AuthResult class.
 */
data class AuthResult(
    val requiresKeySetup: Boolean,
    val isNewUser: Boolean,
    val nextStep: String? = null,
    val message: String? = null
)

/**
 * Auth service — replaces Dart auth_service.dart.
 *
 * Handles all authentication API calls to the Phoenix backend:
 *   • identify (email/phone detection)
 *   • verify_password (login or signup)
 *   • verify_otp (6-digit code verification)
 *   • resend_otp
 *   • send_phone_otp (phone users → email collection)
 *   • forgot_password / reset_password
 *   • Google Sign-In
 *   • Token refresh
 *   • Profile management
 */
class AuthService(
    private val sessionManager: SessionManager
) {
    companion object {
        private const val TAG = "AuthService"
    }

    private val client = OkHttpClient.Builder()
        .connectTimeout(15, java.util.concurrent.TimeUnit.SECONDS)
        .readTimeout(15, java.util.concurrent.TimeUnit.SECONDS)
        .build()

    private val jsonMediaType = "application/json; charset=utf-8".toMediaType()

    private val baseUrl: String get() = AppConfig.apiBaseUrl

    // ─── Step 1: Identify ────────────────────────────────────────

    /**
     * Detect if email/phone exists and get next auth step.
     * Returns map with: temp_token, next_step, message
     */
    suspend fun identify(identifier: String): Map<String, Any?> {
        val body = JSONObject().apply {
            put("identifier", identifier)
        }
        val response = post("$baseUrl/api/v1/auth/identify", body)

        if (response.code == 200) {
            return JSONObject(response.body!!.string()).toMap()
        } else if (response.code == 429) {
            throw IOException("Too many attempts. Please try again later.")
        } else {
            val error = try {
                JSONObject(response.body!!.string()).optString("error", "Identification failed")
            } catch (_: Exception) { "Identification failed" }
            throw IOException(error)
        }
    }

    // ─── Step 2a: Verify Password ────────────────────────────────

    /**
     * Verify password for login (existing user) or signup (new user).
     */
    suspend fun verifyPassword(tempToken: String, password: String): AuthResult {
        val body = JSONObject().apply {
            put("temp_token", tempToken)
            put("password", password)
        }
        val response = post("$baseUrl/api/v1/auth/verify_password", body)

        if (response.code == 200) {
            val data = JSONObject(response.body!!.string())

            // Check if more steps are needed (OTP or email collection)
            val nextStep = data.optString("next_step", "")
            if (nextStep == "otp_verify" || nextStep == "email_collect") {
                return AuthResult(
                    requiresKeySetup = false,
                    isNewUser = true,
                    nextStep = nextStep,
                    message = data.optString("message", null)
                )
            }

            // Full login success — save session
            sessionManager.saveSession(
                token = data.getString("token"),
                userId = data.getString("user_id"),
                deviceId = data.getString("device_id"),
                refreshToken = data.optString("refresh_token", null)
            )
            return AuthResult(
                requiresKeySetup = data.optBoolean("requires_key_setup", false),
                isNewUser = data.optBoolean("is_new_user", false)
            )
        } else if (response.code == 422) {
            val data = JSONObject(response.body!!.string())
            val details = try {
                val arr = data.optJSONArray("details")
                if (arr != null) {
                    (0 until arr.length()).joinToString(", ") { arr.getString(it) }
                } else "Password too weak"
            } catch (_: Exception) { "Password too weak" }
            throw IOException(details)
        } else {
            val error = try {
                JSONObject(response.body!!.string()).optString("error", "Invalid credentials")
            } catch (_: Exception) { "Invalid credentials" }
            throw IOException(error)
        }
    }

    // ─── Step 2b: Verify OTP ─────────────────────────────────────

    /**
     * Verify 6-digit OTP code.
     */
    suspend fun verifyOtp(tempToken: String, otp: String): AuthResult {
        val body = JSONObject().apply {
            put("temp_token", tempToken)
            put("otp", otp)
        }
        val response = post("$baseUrl/api/v1/auth/verify_otp", body)

        if (response.code == 200) {
            val data = JSONObject(response.body!!.string())
            sessionManager.saveSession(
                token = data.getString("token"),
                userId = data.getString("user_id"),
                deviceId = data.getString("device_id"),
                refreshToken = data.optString("refresh_token", null)
            )
            return AuthResult(
                requiresKeySetup = data.optBoolean("requires_key_setup", false),
                isNewUser = data.optBoolean("is_new_user", false)
            )
        } else {
            val error = try {
                JSONObject(response.body!!.string()).optString("error", "Invalid OTP")
            } catch (_: Exception) { "Invalid OTP" }
            throw IOException(error)
        }
    }

    // ─── Resend OTP ──────────────────────────────────────────────

    suspend fun resendOtp(tempToken: String, email: String) {
        val body = JSONObject().apply {
            put("temp_token", tempToken)
            put("email", email)
        }
        val response = post("$baseUrl/api/v1/auth/resend_otp", body)

        if (response.code == 429) {
            throw IOException("Please wait before requesting a new code.")
        } else if (response.code != 200) {
            val error = try {
                JSONObject(response.body!!.string()).optString("error", "Failed to resend")
            } catch (_: Exception) { "Failed to resend" }
            throw IOException(error)
        }
    }

    // ─── Phone OTP (email collection for phone signups) ──────────

    suspend fun sendPhoneOtp(tempToken: String, email: String): AuthResult {
        val body = JSONObject().apply {
            put("temp_token", tempToken)
            put("email", email)
        }
        val response = post("$baseUrl/api/v1/auth/send_phone_otp", body)

        if (response.code == 200) {
            val data = JSONObject(response.body!!.string())
            return AuthResult(
                requiresKeySetup = false,
                isNewUser = true,
                nextStep = data.optString("next_step", null),
                message = data.optString("message", null)
            )
        } else if (response.code == 422) {
            val error = try {
                JSONObject(response.body!!.string()).optString("error", "Invalid email")
            } catch (_: Exception) { "Invalid email" }
            throw IOException(error)
        } else {
            val error = try {
                JSONObject(response.body!!.string()).optString("error", "Failed to send verification")
            } catch (_: Exception) { "Failed to send verification" }
            throw IOException(error)
        }
    }

    // ─── Forgot Password ─────────────────────────────────────────

    suspend fun forgotPassword(email: String) {
        val body = JSONObject().apply {
            put("email", email)
        }
        post("$baseUrl/api/v1/auth/forgot_password", body)
        // Always succeeds (anti-enumeration)
    }

    // ─── Reset Password ──────────────────────────────────────────

    suspend fun resetPassword(email: String, code: String, newPassword: String) {
        val body = JSONObject().apply {
            put("email", email)
            put("code", code)
            put("new_password", newPassword)
        }
        val response = post("$baseUrl/api/v1/auth/reset_password", body)

        if (response.code == 422) {
            val data = JSONObject(response.body!!.string())
            val details = try {
                val arr = data.optJSONArray("details")
                if (arr != null) {
                    (0 until arr.length()).joinToString(", ") { arr.getString(it) }
                } else "Password too weak"
            } catch (_: Exception) { "Password too weak" }
            throw IOException(details)
        } else if (response.code != 200) {
            val error = try {
                JSONObject(response.body!!.string()).optString("error", "Reset failed")
            } catch (_: Exception) { "Reset failed" }
            throw IOException(error)
        }
    }

    // ─── Google Sign-In ──────────────────────────────────────────

    /**
     * Authenticate with Google ID token.
     * Called after Google Credential Manager returns the ID token.
     */
    suspend fun googleAuth(idToken: String): AuthResult {
        val body = JSONObject().apply {
            put("id_token", idToken)
        }
        val response = post("$baseUrl/api/v1/auth/google", body)

        if (response.code == 200) {
            val data = JSONObject(response.body!!.string())
            sessionManager.saveSession(
                token = data.getString("token"),
                userId = data.getString("user_id"),
                deviceId = data.getString("device_id"),
                refreshToken = data.optString("refresh_token", null)
            )
            return AuthResult(
                requiresKeySetup = data.optBoolean("requires_key_setup", false),
                isNewUser = data.optBoolean("is_new_user", false)
            )
        } else {
            val error = try {
                JSONObject(response.body!!.string()).optString("error", "Google login failed")
            } catch (_: Exception) { "Google login failed" }
            throw IOException(error)
        }
    }

    // ─── Token Refresh ───────────────────────────────────────────

    /**
     * Silently refresh the access token using the stored refresh token.
     * Returns true if refresh was successful.
     */
    suspend fun refreshToken(): Boolean {
        val storedRefreshToken = sessionManager.getRefreshToken() ?: return false

        val body = JSONObject().apply {
            put("refresh_token", storedRefreshToken)
        }

        return try {
            val response = post("$baseUrl/api/v1/auth/refresh", body)
            if (response.code == 200) {
                val data = JSONObject(response.body!!.string())
                sessionManager.updateTokens(
                    token = data.getString("token"),
                    refreshToken = data.getString("refresh_token")
                )
                true
            } else {
                false
            }
        } catch (e: Exception) {
            Log.e(TAG, "Token refresh failed: ${e.message}")
            false
        }
    }

    // ─── Profile Management ──────────────────────────────────────

    /**
     * Update user profile (name and avatar).
     */
    suspend fun updateProfile(
        firstName: String,
        lastName: String? = null,
        imagePath: String? = null
    ) {
        val token = sessionManager.getToken() ?: throw IOException("Not authenticated")

        val builder = MultipartBody.Builder().setType(MultipartBody.FORM)
        builder.addFormDataPart("first_name", firstName)
        if (!lastName.isNullOrEmpty()) {
            builder.addFormDataPart("last_name", lastName)
        }

        if (!imagePath.isNullOrEmpty()) {
            val file = java.io.File(imagePath)
            if (file.exists()) {
                val mediaType = "image/jpeg".toMediaType()
                builder.addFormDataPart(
                    "avatar",
                    file.name,
                    file.readBytes().toRequestBody(mediaType)
                )
            }
        }

        val request = Request.Builder()
            .url("$baseUrl/api/v1/account/profile")
            .addHeader("Authorization", "Bearer $token")
            .put(builder.build())
            .build()

        val response = execute(request)
        if (response.code == 401) {
            // Try refresh
            if (refreshToken()) {
                val newToken = sessionManager.getToken() ?: throw IOException("Not authenticated")
                val retryRequest = request.newBuilder()
                    .header("Authorization", "Bearer $newToken")
                    .build()
                val retryResponse = execute(retryRequest)
                if (retryResponse.code != 200) {
                    throw IOException("Failed to update profile")
                }
            }
        } else if (response.code != 200) {
            val error = try {
                JSONObject(response.body!!.string()).optString("error", "Failed to update profile")
            } catch (_: Exception) { "Failed to update profile" }
            throw IOException(error)
        }
    }

    /**
     * Set username and optional PIN.
     */
    suspend fun setUsername(username: String, pin: String? = null) {
        val body = JSONObject().apply {
            put("username", username)
            if (!pin.isNullOrEmpty()) {
                put("pin", pin)
            }
        }
        val response = authenticatedPut("$baseUrl/api/v1/account/username", body)

        if (response.code != 200) {
            val errorBody = try {
                JSONObject(response.body!!.string())
            } catch (_: Exception) { JSONObject() }

            val errors = errorBody.optJSONObject("errors")
            if (errors != null && errors.has("username")) {
                val msgs = errors.optJSONArray("username")
                if (msgs != null && msgs.length() > 0) {
                    throw IOException("Username ${msgs.getString(0)}")
                }
            }
            throw IOException(errorBody.optString("error", "Failed to set username"))
        }
    }

    /**
     * Check if a username is available.
     */
    suspend fun checkUsername(username: String): Boolean {
        return try {
            val encoded = java.net.URLEncoder.encode(username, "UTF-8")
            val request = Request.Builder()
                .url("$baseUrl/api/v1/account/check_username?username=$encoded")
                .get()
                .build()
            val response = execute(request)
            if (response.code == 200) {
                val data = JSONObject(response.body!!.string())
                data.optBoolean("available", false)
            } else {
                false
            }
        } catch (_: Exception) {
            true // Assume available on network error to not block typing
        }
    }

    /**
     * Fetch current user's profile.
     */
    suspend fun getProfile(): Map<String, Any?> {
        val response = authenticatedGet("$baseUrl/api/v1/account/profile")
        if (response.code == 200) {
            return JSONObject(response.body!!.string()).toMap()
        } else {
            throw IOException("Failed to fetch profile")
        }
    }

    // ─── Logout ──────────────────────────────────────────────────

    suspend fun logout() {
        try {
            authenticatedPost("$baseUrl/api/v1/auth/logout", JSONObject())
        } catch (_: Exception) {
            // Best effort
        }
        sessionManager.clearSession()
    }

    // ─── Internal HTTP helpers with auto-refresh ─────────────────

    private suspend fun authenticatedGet(url: String): Response {
        val token = sessionManager.getToken() ?: throw IOException("Not authenticated")
        val request = Request.Builder()
            .url(url)
            .addHeader("Authorization", "Bearer $token")
            .get()
            .build()
        var response = execute(request)

        if (response.code == 401) {
            if (refreshToken()) {
                val newToken = sessionManager.getToken() ?: throw IOException("Not authenticated")
                val retryRequest = request.newBuilder()
                    .header("Authorization", "Bearer $newToken")
                    .build()
                response = execute(retryRequest)
            }
        }
        return response
    }

    private suspend fun authenticatedPost(url: String, body: JSONObject): Response {
        val token = sessionManager.getToken() ?: throw IOException("Not authenticated")
        val requestBody = body.toString().toRequestBody(jsonMediaType)
        val request = Request.Builder()
            .url(url)
            .addHeader("Authorization", "Bearer $token")
            .addHeader("Content-Type", "application/json")
            .post(requestBody)
            .build()
        var response = execute(request)

        if (response.code == 401) {
            if (refreshToken()) {
                val newToken = sessionManager.getToken() ?: throw IOException("Not authenticated")
                val retryRequest = request.newBuilder()
                    .header("Authorization", "Bearer $newToken")
                    .build()
                response = execute(retryRequest)
            }
        }
        return response
    }

    private suspend fun authenticatedPut(url: String, body: JSONObject): Response {
        val token = sessionManager.getToken() ?: throw IOException("Not authenticated")
        val requestBody = body.toString().toRequestBody(jsonMediaType)
        val request = Request.Builder()
            .url(url)
            .addHeader("Authorization", "Bearer $token")
            .addHeader("Content-Type", "application/json")
            .put(requestBody)
            .build()
        var response = execute(request)

        if (response.code == 401) {
            if (refreshToken()) {
                val newToken = sessionManager.getToken() ?: throw IOException("Not authenticated")
                val retryRequest = request.newBuilder()
                    .header("Authorization", "Bearer $newToken")
                    .build()
                response = execute(retryRequest)
            }
        }
        return response
    }

    private suspend fun post(url: String, body: JSONObject): Response {
        val requestBody = body.toString().toRequestBody(jsonMediaType)
        val request = Request.Builder()
            .url(url)
            .addHeader("Content-Type", "application/json")
            .post(requestBody)
            .build()
        return execute(request)
    }

    private suspend fun execute(request: Request): Response {
        return suspendCancellableCoroutine { continuation ->
            val call = client.newCall(request)
            continuation.invokeOnCancellation { call.cancel() }
            call.enqueue(object : Callback {
                override fun onResponse(call: Call, response: Response) {
                    continuation.resume(response)
                }
                override fun onFailure(call: Call, e: IOException) {
                    continuation.resumeWithException(e)
                }
            })
        }
    }
}
